### PR TITLE
Fixes a crash when selecting a different process

### PR DESCRIPTION
**Pull request by florian-kuebler**
This bug occurred when there were symbols loaded and functions hooked.
In this case the process gets updated internally and the UI tries to
render the function window before its `OnDataChanged` event arrived,
as `UpdateModules` fires the refresh callbacks at the very end initially
spawned from another thread.

This fix pospones the setting of the global variable GTargetProcess
until all data is ready to propagete the UI update (in particular the
modules list). So if the UI redraws before the UpdateModules function
propagates the data update, there will be no inconsistent state and
thus no crash. As UpdateModules now also sets the process, the method
is also renamed.

This also slightly refactors the Capture::SetTargetProcess by making
it non-ref parameter (as it is a sink) and move the process there and
further rename the parameter.

Bug: http://b/163786513
Test: On Linux: Select a process, load some modules and hook functions.
        Then select another process. The previous crash is gone now.

### CODE REVIEW

File: OrbitGl/App.cpp [hunk 1]
```diff
@@ -873,6 +873,7 @@ void OrbitApp::OnProcessSelected(int32_t pid) {
   std::shared_ptr<Process> process = FindProcessByPid(pid);
   CHECK(process != nullptr);
   Capture::SetTargetProcess(std::move(process));
+  FireRefreshCallbacks();
```
Comment (pierricgimmig) [inline OrbitGl/App.cpp:876]: Could this be moved below UpdateModuleList(pid); ?

Comment (florian-kuebler) [reply OrbitGl/App.cpp:876]: This is actually already in UpdateModuleList(pid) at the very end. However, this method spawns a task on another thread (from the main thread) which then again spawns a task on the main thread. In this last task the FireRefreshCallbacks is called at the very end. And the UI seems to redraw the FunctionsView before that call arrives, and thus has inconsistent data (new process (including functions) and old list of indices of functions to draw), which finally leads to the crash.

I am not sure if there is a better way to solve this without blocking the UI for longer time.
Steps to reproduce and further info are in the bug (b/163786513). I would leave this open for now if you want to look a bit further into this and will merge that change for now.

**Comment by florian-kuebler** [comment]
> Hi Florian, thanks for the fix. Your description mentions that this is a naive fix, is there a better fix that we could implement? If so we should open a bug. I'm ok with submitting as is as it fixes a crash, but we should possibly look further into it. Would you have repro steps or a callstack? Thanks.

Thanks for looking into this. See my [comment above](https://github.com/google/orbit/pull/997#discussion_r469718802) for additional information.

File: OrbitGl/App.cpp [hunk 2]
```diff
@@ -873,6 +873,7 @@ void OrbitApp::OnProcessSelected(int32_t pid) {
   std::shared_ptr<Process> process = FindProcessByPid(pid);
   CHECK(process != nullptr);
   Capture::SetTargetProcess(std::move(process));
```
Comment (dimitry-) [inline OrbitGl/App.cpp:875]: Should we set taget process after modules have successfully been updated?

Comment (florian-kuebler) [reply OrbitGl/App.cpp:875]: I guess UpdateModules uses the process. But we can pass it as a param. Maybe this is better. I'll look into this again.

File: OrbitGl/App.cpp (hunk 1 above)
Comment (antonrohr) [reply OrbitGl/App.cpp:876]: I am pretty indifferent between this line as a fix or dimitrys idea in the comment below. I think both should work. If you go with this, maybe `FireRefreshCallbacks(DataViewType::FUNCTIONS);` is enough?

File: OrbitGl/App.cpp (hunk 2 above)
Comment (antonrohr) [reply OrbitGl/App.cpp:875]: UpdateModuleList(pid) also uses FindProcessById(pid). You can probably leave it as is. If I understood the problem, the problem is having a delay between SetTargetProcess and FireRefreshCallbacks.

Comment (florian-kuebler) [reply OrbitGl/App.cpp:875]: Yep you were right. I implemented it this way. Seems to be the proper solution for the issue.

**Comment by florian-kuebler** [comment]
Thanks for the reviews. I changed the fix in order to set the process in UpdateModules at the end.

File: OrbitGl/App.cpp (hunk 1 above)
Comment (florian-kuebler) [reply OrbitGl/App.cpp:876]: I also tried this, but I am not sure, if there were similar bugs in the other views (even though there shouldn't). Anyways, went for the other solution.

**Comment by florian-kuebler** [comment]
Forgot to mention, also updated the commit message.

File: OrbitGl/App.cpp [hunk 3]
```diff
@@ -873,6 +873,7 @@ void OrbitApp::OnProcessSelected(int32_t pid) {
   std::shared_ptr<Process> process = FindProcessByPid(pid);
```
Comment (dimitry-) [inline OrbitGl/App.cpp:873]: ok, one small thing about commit message

Move link to the bug into Bug: stanza

At the end of commit message add
Bug: b/42
Test: <describe how you tested it>

Comment (florian-kuebler) [reply OrbitGl/App.cpp]: Okay done.

File: OrbitCore/Capture.cpp [hunk 4]
```diff
@@ -31,10 +31,10 @@ std::shared_ptr<PresetFile> Capture::GSessionPresets = nullptr;
 
 void Capture::Init() { GTargetProcess = std::make_shared<Process>(); }
```
Comment (dimitry-) [inline OrbitCore/Capture.cpp:32]: nit: you can remove brackets around (b/42) and if you add http:// before you will be able to open it from console while looking at git log (as it is now you will need to copy and paste it)

Comment (florian-kuebler) [reply OrbitCore/Capture.cpp:32]: Sure. Done!

Comment (dimitry-) [reply OrbitCore/Capture.cpp:32]: ehm, test does not describe how to test this change. I will not be able to reproduce it.

File: OrbitGl/App.cpp [hunk 5]
```diff
@@ -860,7 +861,12 @@ void OrbitApp::UpdateModuleList(int32_t pid) {
         LoadModulesFromPreset(process, preset);
         Capture::GSessionPresets = nullptr;
       }
-      // To this point ----------------------------------
```
Comment (dimitry-) [inline OrbitGl/App.cpp:863]: This was a terminator for TODO comment :)

Comment (florian-kuebler) [reply OrbitGl/App.cpp:863]: Ups.  Sorry. Readded it again.

**Comment by florian-kuebler** [comment]
Added proper "Test:" description.

File: OrbitCore/Capture.cpp [hunk 6]
```diff
@@ -31,10 +31,10 @@ std::shared_ptr<PresetFile> Capture::GSessionPresets = nullptr;
...
 
-void Capture::SetTargetProcess(const std::shared_ptr<Process>& a_Process) {
-  if (a_Process != GTargetProcess) {
-    GTargetProcess = a_Process;
-    GSamplingProfiler = std::make_shared<SamplingProfiler>(a_Process);
+void Capture::SetTargetProcess(std::shared_ptr<Process> process) {
+  if (process != GTargetProcess) {
+    GTargetProcess = std::move(process);
+    GSamplingProfiler = std::make_shared<SamplingProfiler>(process);
```
Comment (antonrohr) [inline OrbitCore/Capture.cpp:37]: Don't use `process` after move

File: OrbitGl/App.cpp [hunk 7]
```diff
@@ -870,12 +878,8 @@ void OrbitApp::UpdateModuleList(int32_t pid) {
 void OrbitApp::OnProcessSelected(int32_t pid) {
```
Comment (antonrohr) [inline OrbitGl/App.cpp:878]: Nit: This function can be removed

File: OrbitCore/Capture.cpp (hunk 6 above)
Comment (florian-kuebler) [reply OrbitCore/Capture.cpp:36]: Got me! Thanks!

File: OrbitGl/App.cpp (hunk 7 above)
Comment (florian-kuebler) [reply OrbitGl/App.cpp]: I removed the function.

**Pull request merged**